Improve Street controls 1

gliegard opened the pull request:
## Description
improve StreetControls : add click and go, surface helpers, click and look at, tweens animations :
- Mouse click on ground detection, and associated callback, that go to the clicked position
- Mouse click on other object (not ground), and associated callback, that makes camera look at the position
- Surfaces as helpers for the end user :
  - Circle displayed when user moves the mouse on the ground
  - Rectangle displayed when user moves the mouse on other objects (like walls of a building)
- Tweens animations :
  - camera can move to another position smoothly
  - camera can look at a position smoothly in an animation

## Motivation and Context
User should have a good experience using immersive example

## Screenshots (if appropriate)
![helper_wall](https://user-images.githubusercontent.com/15127065/51912450-acc2d780-23d4-11e9-861f-1fcf81813d58.png)
![helper_ground](https://user-images.githubusercontent.com/15127065/51912459-b3e9e580-23d4-11e9-864b-198c81cb57cd.png)

gliegard committed:
feat(StreetControls): add click and go, surface helpers, click and look at, tweens animations :
- Mouse click on ground detection, and associated callback, that go to the clicked position
- Mouse click on other object (not ground), and associated callback, that makes camera look at the position
- Surfaces as helpers for the end user :
  - Circle displayed when user moves the mouse on the ground
  - Rectangle displayed when user moves the mouse on other objects (like walls of a building)
- Tweens animations :
  - camera can move to another position smoothly
  - camera can look at a position smoothly in an animation

gliegard committed:
feat(TileMesh): add a boolean to know if it's a TileMesh or not, without doing an instanceof

gliegard committed:
fix(OrientedImageCamera): increase default far from 1000 to 10 000.
Before this commit, orientedImageCamera, has a default far set to 1000,
A side effect is that OrientedImageMaterial was only able to project texture to a distance of 1000,
because OrientedImageMaterial uses OrientedImageCamera to project textures.
Now, OrientedImageCamera default far set to 10 000, so OrientedImageMaterial projects texture to 10 km distance.

gchoqueux commented on src/Core/TileMesh.js:
```diff
@@ -34,6 +34,7 @@ class TileMesh extends THREE.Mesh {
         this.rotationAutoUpdate = false;
 
         this.layerUpdateState = {};
+        this.isTileMesh = true;
```
Could you replace the condition `(node instanceof TileMesh)` by (this.isTileMesh)

gliegard replied:
ok

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -1,17 +1,23 @@
+/* globals window */
```
You can remove it

gliegard replied:
ok

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -1,17 +1,23 @@
  * It is used to simulate a street view.
  * It stores a currentPosition and nextPosition, and do a camera traveling to go to next position.
- *
+ * It also manages picking on the ground and on other object, like building.
```
:+1:

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -1,17 +1,23 @@
+/* globals window */
+import { MAIN_LOOP_EVENTS } from 'Core/MainLoop';
```
Can you move this import after three ?

gliegard replied:
ok

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +79,143 @@ class StreetControls extends FirstPersonControls {
+        }
+    }
+
+    pick(event, pickGround = () => {}, pickObject =  () => {}) {
```
Mind your spaces

gliegard replied:
oki

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +79,143 @@ class StreetControls extends FirstPersonControls {
+        const geomCircle = new THREE.CircleGeometry(1, 32);
+        this._circleMesh = new THREE.Mesh(geomCircle, material);
+
+        this.view.scene.add(this._rectangleMesh);
```
Why adding it if you add it below in `this.surfacesNode`, which is added to the scene ?

gliegard replied:
yep it's changed

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +79,143 @@ class StreetControls extends FirstPersonControls {
+    firstPickedObject(event) {
+        // basic threeJs pick with raycaster
+        const normalized = this.view.viewToNormalizedCoords(event);
+        this.raycaster.setFromCamera(normalized.clone(), this.camera);
```
Could you use picking from itowns?

gliegard replied:
No, because existing picking methods are picking in a specific layer, and I want this StreetControls to be generic : it doesn't know the other layers, it picks all the scene, and will work with every layers (3dTiles, WFS buildings...).

gliegard replied:
Finally, I use existing picking method !! :)

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -102,19 +239,123 @@ class StreetControls extends FirstPersonControls {
+    }
+
+    /**
+     * Method called when user click on oject that is not the ground.</br>
```
Pro-tip: you can avoid using `<br/>` by adding another breakline, thus also making the code more readable.

gliegard replied:
Adding a newline in the comment don't work in the doc.

zarov replied:
Weird, I thought it would. Thanks !

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -102,19 +239,123 @@ class StreetControls extends FirstPersonControls {
+        const distance = this.camera.position.distanceTo(position);
+
+        // can't click on a wall that is at 1km distance.
+        if (distance < 1000) {
```
This value is very arbitrary, could you set it in a variable in the constructor ?

gliegard replied:
ok

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -102,19 +239,123 @@ class StreetControls extends FirstPersonControls {
+        this.start = this.camera.clone();
+        this.end = this.camera.clone();
+        this.end.lookAt(position);
+        const factor = { t: 0 };
```
No need for a `const`

gliegard replied:
I don't understand what you mean. The object 'factor' is constant, even if it's content 'factor.t' is varying.

zarov replied:
Sorry, I meant that you could directly write `new TWEEN.Tween({ t: 0 }, this.tweenGroup)` below.

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -63,6 +57,14 @@ class StreetControls extends FirstPersonControls {
+        this.tweenGroup = new TWEEN.Group();
+
+        // init surfaces used as helper for end user
+        this.initSurfaces();
```
if `initSurfaces` is used to instantiate the helpers, you could make a private fonction, like this:
```js
initSurfacesHelper(this);
```

gliegard replied:
yes good idea

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +79,143 @@ class StreetControls extends FirstPersonControls {
+        // - skip the two surfaces (rectangle and circle)
+        // - skip all the object without face, because we use the face to manage surface angle, so it's needed.
+        let i = 0;
+        while (i < intersects.length &&
```
With picking itowns, you can filter by layer. this should avoid this loop.

gliegard replied:
I don't manage to make it work with existing picking method from iTowns, because I need to pick the TileMesh object, to get the face and the normal. So I prefer using threeJs picking.

gliegard replied:
Finally, I use existing picking method !! :)

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -63,6 +57,14 @@ class StreetControls extends FirstPersonControls {
+        // Raycaster used to pick objects
+        this.raycaster = new THREE.Raycaster();
+        // Tween is used to make smooth animations
+        this.tweenGroup = new TWEEN.Group();
```
Declare `this.tweens` (which should be `this.tween`) here. Side note: why do you need a group and a tween ? (I know nothing about this library)

gliegard replied:
ok. side answer: First I try without a Group but I didn't manage to make it work.

zarov replied:
Ok, I'll also take a look at it to see if we can make it work

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +79,143 @@ class StreetControls extends FirstPersonControls {
+    pick(event, pickGround = () => {}, pickObject =  () => {}) {
+        // specific method to pick on the elevation layer at the good elevation, even if elevation layer is done in the shader
+        const mouse = this.view.eventToViewCoords(event);
+        const target = new THREE.Vector3();
```
could you move out `target` from this function .

gliegard replied:
ok

gchoqueux replied:
```js
 pick(event, pickGround = () => {}, pickObject = () => {}, target = new THREE.Vector3())
```
 you don't use `target` apart from this function, you could remove it

gliegard replied:
ok, I remove it (this method has been refactored during the review)

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +79,143 @@ class StreetControls extends FirstPersonControls {
+    onClick() {
+        // here we detect mouse click when mouse is not moving,
+        // to avoid interfering with mouse drag movement, managed by FirstPersonControls
+        this.maybeClickWithoutMoving = true;
```
may be use state machine would be more clear. WDYT?

gliegard replied:
yep, I replace it with a boolean '_stateOnMouseDrag' , it's clearer.

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +79,143 @@ class StreetControls extends FirstPersonControls {
+        // basic threeJs pick with raycaster
+        const normalized = this.view.viewToNormalizedCoords(event);
+        this.raycaster.setFromCamera(normalized.clone(), this.camera);
+        const intersects = this.raycaster.intersectObjects(this.view.scene.children, true);
```
This will be very very heavy if you are inside a building with a complex mesh: we need a better solution. Now, I'm not going to block this PR for this, but I would appreciate if you already have some insights on how to fix this. 

Anyway, please add a warning concerning this problem for now.

gliegard replied:
Finally, I use existing picking method, and pick on specific layers !!

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +79,143 @@ class StreetControls extends FirstPersonControls {
+
+    // hide both rectangle and circle helpers
+    hideSurfaces() {
+        this._circleMesh.visible = false;
```
Could you use a `THREE.Group` to manage the `surfaceHelpers`. You could write only `surfaceHelpers.visible = false`.

gliegard replied:
No because, on the other hand, when I will want to make only one surface visible, I'll have to make the Group visible again, so this idea remove on line of code somewhere, but make me add another line at other place. So I think it's not a good idea.

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +79,143 @@ class StreetControls extends FirstPersonControls {
+    onClick() {
+        // here we detect mouse click when mouse is not moving,
+        // to avoid interfering with mouse drag movement, managed by FirstPersonControls
+        this.maybeClickWithoutMoving = true;
```
This seems fishy: can you explain the interference you got ?

gliegard replied:
I've already changed it.

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +79,143 @@ class StreetControls extends FirstPersonControls {
+            opacity: 0.5 });
+
+        const geomPlane = new THREE.PlaneGeometry(4, 2, 1);
+        this._rectangleMesh = new THREE.Mesh(geomPlane, material);
```
It would be nice to be able to customise those meshes

gliegard replied:
You don't like these ones ?
;)
Yes good idea, I change that

zarov replied:
I like them, but we never know, someone may want a heart and a flower ;) More seriously, we could imagine a sort of game where you are moving a spider for example, and then replace the two meshes by a spider.

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -102,19 +239,123 @@ class StreetControls extends FirstPersonControls {
+     *
+     * @param {THREE.Vector3} position - The position
+     */
+    onClickOnGround(position) {
```
these functions don't need to be exposed.

gliegard replied:
As I write it on the doc 

"Note that this funtion contains default values that can be overrided, by overriding this function."

That means these methods are the default behavior , and it can be overrided to do specific behavior. So it's designed to be exposed.

gchoqueux replied:
I can't overload them. How are you doing?

gliegard replied:
```
            class myControl extends itowns.StreetControls {
                onClickOnGround(position) {
                    console.log('user click on ground !');
                }
            }
```

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -102,19 +239,123 @@ class StreetControls extends FirstPersonControls {
+     */
+    onClickOnGround(position) {
+        // compute a new position to go : 4 meter up from the picked position on the ground
+        const up = position.clone().normalize();
```
does it work with the planar?

gliegard replied:
Not yet, but I plan to add another example in planar. At this moment I will make all this work in planar view.

gchoqueux replied:
Could you create a issue and add comment in the documentation?

gliegard replied:
yes #1036

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -102,19 +239,123 @@ class StreetControls extends FirstPersonControls {
+    onClickOnGround(position) {
+        // compute a new position to go : 4 meter up from the picked position on the ground
+        const up = position.clone().normalize();
+        position.add(up.multiplyScalar(4));
```
it's necessary to multiply by 4?

gliegard replied:
It's a default value for the given example. The goal is to move 4 meter up to the ground. It can be overrided, by overriding the method.

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -102,19 +239,123 @@ class StreetControls extends FirstPersonControls {
+     * @param { THREE.Vector3 }  position - Position to look at
+     * @param { number } time - Time in millisecond
+     */
+    animateCameraLookAt(position, time) {
```
it would be interesting to have `animateCameraLookAt` and `moveCameraTo` in `CameraUtils`.

gchoqueux replied:
could you open a issue? to move this part in `cameraUtils` in the futur?

gliegard replied:
yes, i've updated #1036 !

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -102,19 +239,123 @@ class StreetControls extends FirstPersonControls {
+    onMouseOut() {
+        this.maybeClickWithoutMoving = false;
+        this.onMouseUp();
+        this.stopTweensAnimations();
```
why you stop the animation.

gliegard replied:
Finally I don't stop animations anymore.

gchoqueux reviewed (changes requested):
Nice improvement

zarov commented on src/Core/TileMesh.js:
```diff
@@ -34,6 +34,7 @@ class TileMesh extends THREE.Mesh {
         this.rotationAutoUpdate = false;
 
         this.layerUpdateState = {};
+        this.isTileMesh = true;
```
Do the appropriate change in `convertToTile` and `Picking` please.

gliegard replied:
yep

zarov reviewed (changes requested):
Could you add an explanation in the message of your `fix` commit ?

Anyway, very happy to see something like this, though it needs a bit more work. Thanks a lot !

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -1,17 +1,58 @@
+    surfaces.add(rectangle);
+    surfaces.add(circle);
+
+    // to acces specific surface, we also add it to the surfaces object
```
I'd prefer to use [some functions of Object3D](https://threejs.org/docs/#api/en/core/Object3D.getObjectById) instead of adding properties.

gliegard replied:
The problem is I don't know the Id given by ThreeJs, and since we can add a custom Mesh, I can't sure of the name of the Mesh.
I change the code to add these meshes as properties of this StreetControls.

zarov replied:
Good point, nvm then

gliegard committed:
example(immersive): add debug UI, and change the name of 'WFS buildings' layer to 'Buildings', and increase background radius from 80 to 1 200

gchoqueux commented on examples/immersive_paris.html:
```diff
@@ -162,6 +162,9 @@
             var menuGlobe = new GuiTools('menuDiv');
             menuGlobe.view = view;
 
+            var d = new debug.Debug(view, menuGlobe.gui);
```
Change opacity building doesn't work, may be projection material hasn't opacity propertie, no?

gliegard replied:
Yes, thanks, I've updated OrientedImageMaterial and shaders. Now it supports opacity.

gchoqueux commented on examples/immersive_paris.html:
```diff
@@ -171,6 +174,15 @@
                 // set camera to current panoramic
                 view.controls.setCameraToCurrentPosition();
                 view.notifyChange(view.camera.camera3D);
+
```
Could you use `view.getLayerById('Building')` instead of `for loop`?

gliegard replied:
yes, I've removed the loop. Finally I add the debug UI before,  when I add the layer.

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -1,17 +1,41 @@
-    const dampingProgress = 1 - Math.pow((1 - (Math.sin((progress - 0.5) * Math.PI) * 0.5 + 0.5)), 2);
-    root.camera.position.lerpVectors(root.positionFrom, root.positionTo, dampingProgress);
+const material = new THREE.MeshBasicMaterial({
+    side: THREE.DoubleSide,
```
Why is double sided?

gliegard replied:
Ok, I fix one surface orientation, so I don't need DoubleSide anymore. Thanks !

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +113,74 @@ class StreetControls extends FirstPersonControls {
 
+    pick(event, pickGround = () => {}, pickObject = () => {}, target = new THREE.Vector3()) {
+        // get real distance to ground, with a specific method to pick on the elevation layer
+        this.view.getPickingPositionFromDepth(this.view.eventToViewCoords(event), target);
```
It would be necessary to open an issue to add the distance calculation in `Picking.pickTilesAt`.

gliegard replied:
ok #1037

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +113,74 @@ class StreetControls extends FirstPersonControls {
+
+        // to detect pick on building, compare first picked building distance to ground distance
+        if (intersectsBuildings.length && intersectsBuildings[0].distance < distanceToGround) { // pick buildings
+            this.updateSurface(intersectsBuildings[0].point, intersectsBuildings[0].face.normal, this.surfaceWall, this.surfaceGround);
```
There are certain functions that do not need to be exposed (`updateSurface, onMouseUp`...)

gliegard replied:
Yes I moved some of them

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -102,19 +204,114 @@ class StreetControls extends FirstPersonControls {
+        this.view.removeFrameRequester(MAIN_LOOP_EVENTS.BEFORE_RENDER, this.animationFrameRequester);
+    }
+
+    stopTweensAnimations() {
```
could you rename in `stopAnimation`? the user doesn't know it's `tween`.
could add `this.view.removeFrameRequester(MAIN_LOOP_EVENTS.BEFORE_RENDER, this.animationFrameRequester);` in `stopTweensAnimations`? 

or merge `onComplete` and `stopTweensAnimations`...

gliegard replied:
yes, I've renamed to stopAnimation. And I 've remove onComplete because it's not need to be exposed.

gchoqueux replied:
`this.view.removeFrameRequester` isn't removed in all cases

gliegard replied:
yes, thanks, updated

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -63,6 +79,26 @@ class StreetControls extends FirstPersonControls {
         this.keySetCameraToCurrentPosition = 65;
+
+        // Raycaster used to pick objects
+        this.raycaster = new THREE.Raycaster();
```
could you remove `raycaster`, it's useless

gliegard replied:
yep

gchoqueux commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +113,74 @@ class StreetControls extends FirstPersonControls {
+        const distanceToGround = this.camera.position.distanceTo(target);
+
+        const intersectsBuildings = this.view.pickObjectsAt(event, 0, this.buildingsLayer);
+        const intersectsGround = this.view.pickObjectsAt(event, 0, this.view.tileLayer);
```
you could do the ground picking only if there's not picking with the building

gliegard replied:
Yes

gliegard committed:
fix(OrientedImageMaterial): it supports opacity.
To see an example, look at immersive_paris.html example, you can change building layer transparency in the debug UI.

zarov commented on src/Controls/StreetControls.js:
```diff
@@ -77,8 +160,39 @@ class StreetControls extends FirstPersonControls {
+        }
+    }
+
+    onMouseMove(event) {
```
I have an error when moving the mouse before the layer is ready. Could you add a condition like `if (!layer.ready)` ? Or something that fixes the error.

gliegard replied:
Yes, thanks, I 've added a bugfix in View

zarov reviewed (commented):
Overall it seems the quality downgraded. When i click on the street to move, it takes one second to start, and then it is not smooth at all.

gliegard committed:
fix(view): picking: Add a log warning when user try to pick on a layer not ready.
Before this commit, picking on a layer, that was not yet ready, thrown an exception.

gchoqueux reviewed (approved):
thanks
